firmware/tools/bitmap_tools/pp_png2hpp.py

```diff
@@ -0,0 +1,217 @@
+	
+	return (out_bmpdata)
+
+def pp_bitmaphpp_data(pngicons_path):
+    count = 0
+    bitmaphpp_data = []
+    for file in os.listdir(pngicons_path):
+        if os.path.isfile(pngicons_path + file) and file.endswith(".png"):
+            bitmaphpp_data.append(convert_png(pngicons_path + file))
+            count += 1
+
+    return bitmaphpp_data
+    
+
+def pp_bitmaphpp_footer():
+    return("""
+} /* namespace ui */
+	
+#endif /*__BITMAP_HPP__*/
+""")
+
+
+def pp_write_bitmaphpp(pngicons_path, hpp_outpath):
+    bitmaphpp_file = []
+
+    bitmaphpp_file.append(pp_bitmaphpp_header())
+    bitmaphpp_file.append("".join(str(x) for x in pp_bitmaphpp_data(pngicons_path)))
+    bitmaphpp_file.append(pp_bitmaphpp_footer())
+    
+    out_file = "".join(str(x) for x in bitmaphpp_file)
+
+    with open(hpp_outpath, "w", encoding="utf-8") as fd:
+        fd.writelines(out_file)
+    
+    print("Find your bitmap.hpp at", out_file)
+
+
+### Convert from a bitmap.hpp file one or all icons in png.
+###########################################################
+
+def parse_bitmaphpp(bitmaphpp_file):
+    ico_pattern = re.compile(r"static constexpr uint8_t bitmap_(.*)_data\[\] = {\n((?:\s+(?:.*)\n)+)};\nstatic constexpr Bitmap bitmap_.*\{\n\s+\{(.*)\},", re.MULTILINE)
+    ico_data = []
+    
+    # read file to buffer, to find multiline regex
+    readfile = open(bitmaphpp_file,'r')
+    buff = readfile.read()
+    readfile.close()
+
+    for match in ico_pattern.finditer(buff):
+        ico_data.append([match.group(1), match.group(2), match.group(3)])
+
+    return (ico_data)
+
+
+def convert_hpp(icon_name,bitmap_array,iconsize_str,png_outdir):
+    iconsize = iconsize_str.split(", ")
+    bitmap_size = (int(iconsize[0]),int(iconsize[1]))
+    bitmap_data=[]
+
+    image_data = np.zeros((bitmap_size[1], bitmap_size[0]), dtype=np.uint8)
+
+    for value in bitmap_array.split(",\n"):
+        if (value):
+            bitmap_data.append(int(value, 0))
+
+    for y in range(bitmap_size[1]):
+        for x in range(bitmap_size[0]):
+            byte_index = (y * bitmap_size[0] + x) // 8
+            bit_index = x % 8
+            # bit_index = 7 - (x % 8)
+            pixel_value = (bitmap_data[byte_index] >> bit_index) & 1
+            image_data[y, x] = pixel_value * 255
+
+    image = Image.fromarray(image_data, 'L')
+    icon_out = os.path.join(png_outdir,icon_name+".png")
+    image.save(icon_out)
+
+
+### Processing
+##############
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument("hpp", help="Path for bitmap.hpp")
+    parser.add_argument("graphics", help="Path of <icon>.png files to convert", default=os.path.join(os.getcwd(),"graphics"))
+    parser.add_argument("--icon","-i", help="Name of the icon to convert")
+    parser.add_argument("--reverse","-r", help="Convert icon from bitmap.hpp to <name>.png", action="store_true")
+    
+    args = parser.parse_args()
+
+    if args.reverse:
```

Comment:
it seems the logic is a little glitched here, you did't pass hpp file path argument to the flow but passed graphics path twice, i think it would cause it override itself, or directly exception

Reply:
this is what i think you could edit to, how do you think?
```
if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument("hpp", help="Path for bitmap.hpp")
    parser.add_argument("graphics", help="Path of <icon>.png files to convert", default=os.path.join(os.getcwd(),"graphics"))
    parser.add_argument("--icon","-i", help="Name of the icon to convert")
    parser.add_argument("--reverse","-r", help="Convert icon from bitmap.hpp to <name>.png", action="store_true")

    args = parser.parse_args()

    if args.reverse:
        print("Reverse: Converting from hpp to png")
        # filter hpp arg is a file
        hpp_file_path = args.hpp
        if not os.path.isfile(hpp_file_path):
            print(f"Error: {hpp_file_path} is not a valid file.")
            sys.exit(1)
        icons = parse_bitmaphpp(hpp_file_path)

        # filter graph arg is a path
        if args.graphics:
            graphics_path = os.path.join(args.graphics, '')
        else:
            graphics_path = os.path.join(os.getcwd(),"graphics", '')
        if not os.path.exists(graphics_path):
             os.makedirs(graphics_path)  # create if not exist

        for icon in icons:
            print("Converting icon", icon[0])
            convert_hpp(icon[0], icon[1], icon[2], graphics_path)
        sys.exit()
    else:
        print("Converting from png to hpp")
        if args.graphics:
            graphics_path = os.path.join(args.graphics, '')
        else:
            graphics_path = os.path.join(os.getcwd(),"graphics", '')
        print("Path", graphics_path, "hpp", args.hpp)
        pp_write_bitmaphpp(graphics_path, args.hpp)
        sys.exit()
```

Reply:
Thanks a lot for your valuable input. Looks much more clean, so I take your code, with a little adjustment to the new icon-handling.